### PR TITLE
ENH: Uniformize the grid&affine across EPI "blips" before TOPUP

### LINKED ISSUE #192: Improve PEPOLAR robustness
Currently, we are assuming all affines match across the EPI fieldmaps in pepolar correction.

A more robust solution would include adding the following steps:
- Estimate a reference image for each phase encoding direction (`StructuralReference`)
- Use one of these images as the reference, and the others as moving images to register.

### PR DESCRIPTION
**Pull request by mgxd**
Closes #192 

Adds a preprocessing workflow for TOPUP's EPI files.

The workflow:
- Takes in a list of EPI files, each pertaining to one phase-encoding (PE) direction.
- Flattens any 4D EPIs into 3D files, and makes a template "blip".
- Applies N4's INU correction.
- Registers all "blips" to the same space.
- Merges the "blips" into a 4D file to be used by TOPUP.

### CODE REVIEW

**Comment by mgxd** [comment]
@oesteban when testing this branch, TOPUP is instantly failing with the following:

<details>
<summary>command</summary>

```
topup --config=/usr/local/miniconda/lib/python3.7/site-packages/sdcflows/data/flirtsch/b02b0.cnf --datain=/scratch/fmriprep_wf/single_subject_voice969_wf/fmap_preproc_wf/wf_auto_00000/topup/template_corrected_merged_encfile.txt --imain=/scratch/fmriprep_wf/single_subject_voice969_wf/fmap_preproc_wf/wf_auto_00000/prepare_blips_wf/merge_blips/template_corrected_merged.nii.gz --out=template_corrected_merged_base --iout=template_corrected_merged_corrected.nii.gz --fout=template_corrected_merged_field.nii.gz --jacout=jac --logout=template_corrected_merged_topup.log --rbmout=xfm --dfout=warpfield
Standard output:

Standard error:

Part of FSL (build 509)
topup

Usage: 
topup --imain=<some 4D image> --datain=<text file> --config=<text file with parameters> --out=my_topup_results


Compulsory arguments (You MUST set one or more of):
	--imain		name of 4D file with images
	--datain	name of text file with PE directions/times

Optional arguments (You may optionally specify one or more of):
	--out		base-name of output files (spline coefficients (Hz) and movement parameters)
	--fout		name of image file with field (Hz)
	--iout		name of 4D image file with unwarped images
	--logout	Name of log-file
	--warpres	(approximate) resolution (in mm) of warp basis for the different sub-sampling levels, default 10
	--subsamp	sub-sampling scheme, default 1
	--fwhm		FWHM (in mm) of gaussian smoothing kernel, default 8
	--config	Name of config file specifying command line arguments
	--miter		Max # of non-linear iterations, default 5
	--lambda	Weight of regularisation, default depending on --ssqlambda and --regmod switches. See user documetation.
	--ssqlambda	If set (=1), lambda is weighted by current ssq, default 1
	--regmod	Model for regularisation of warp-field [membrane_energy bending_energy], default bending_energy
	--estmov	Estimate movements if set, default 1 (true)
	--minmet	Minimisation method 0=Levenberg-Marquardt, 1=Scaled Conjugate Gradient, default 0 (LM)
	--splineorder	Order of spline, 2->Qadratic spline, 3->Cubic spline. Default=3
	--numprec	Precision for representing Hessian, double or float. Default double
	--interp	Image interpolation model, linear or spline. Default spline
	--scale		If set (=1), the images are individually scaled to a common mean, default 0 (false)
	--regrid		If set (=1), the calculations are done in a different grid, default 1 (true)
	-h,--help	display help info
	-v,--verbose	Print diagonostic information while running
	-h,--help	display help info




Return code: 1
```

</details>

It seems it doesn't like the `--config` option, though I'm not sure why (all files look like what we'd expect)

File: sdcflows/workflows/fit/pepolar.py [hunk 1]
```diff
@@ -327,29 +316,49 @@ def init_prepare_blips_wf(*, omp_nthreads=1, orthogonal=True, name="prepare_blip
...
+        niu.IdentityInterface(fields=["epi_files", "metadata"]),
+        name='inputnode'
+    )
+    outputnode = pe.Node(
+        niu.IdentityInterface(fields=["reg_blips", "readout_times"]),
+        name="outputnode"
+    )
 
-    flatten = pe.MapNode(Flatten(), name='flatten')
+    flatten = pe.MapNode(Flatten(), name='flatten', iterfield=["in_data", "in_meta"])
```
Comment (oesteban) [inline sdcflows/workflows/fit/pepolar.py]: This is not going to do what you want it to do.

**Comment by oesteban** [comment]
@mgxd I'll be working on this until the meeting.

**Comment by mgxd** [comment]
> when testing this branch, TOPUP is instantly failing
> It seems it doesn't like the `--config` option, though I'm not sure why (all files look like what we'd expect)

I was able to track this down - turns out it was (like you suggested) a problem with the even/odd number of slices.

**Comment by oesteban** [comment]
Let's split the problem:
- The even/odd number of voxels (although it is weird it worked if passing arguments in the command line)
- The differences in grinding

I've been thinking this through and I'm more convinced that we should use the reference workflow as input for the topup workflow. But, before that can be done, I guess we can just trust that inputs are roughly aligned - TOPUP will run head-motion internally anyways.

So, I'm testing a simplification where we only add one node that ensures all volumes at the input have the same shape/affine at the output.

**Comment by oesteban** [comment]
Seems like some pre-alignment is indeed necessary: https://3047-189236569-gh.circle-artifacts.com/0/tmp/tests/unittests/topup_ds001771/sdcflows/sub-36/figures/sub-36_desc-pepolar_fieldmap.svg

**Comment by oesteban** [comment]
Alignment is only necessary for debug settings as the topup config does not apply it.

**Pull request merged**